### PR TITLE
Canvas Header Stamping

### PR DESCRIPTION
**Pull request by dobosevych**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

### CODE REVIEW

File: celery/app/amqp.py [hunk 1]
```diff
@@ -319,26 +320,29 @@ def as_task_v2(self, task_id, name, args=None, kwargs=None,
...
+            'timelimit': [time_limit, soft_time_limit],
+            'root_id': root_id,
+            'parent_id': parent_id,
+            'argsrepr': argsrepr,
+            'kwargsrepr': kwargsrepr,
+            'origin': origin or anon_nodename(),
+            'ignore_result': ignore_result,
+            'stamps': stamps,
+        }
+        headers.update(options)
```
Comment (thedrow) [inline celery/app/amqp.py:342]: Why not just expand options into headers using **options?

File: celery/app/task.py [hunk 2]
```diff
@@ -784,8 +786,12 @@ def apply(self, args=None, kwargs=None,
                 'routing_key': options.get('routing_key'),
                 'priority': options.get('priority'),
             },
-            'groups': maybe_list(options.get('groups'))
+            'groups': maybe_list(options.get('groups')),
```
Comment (thedrow) [inline celery/app/task.py:789]: Why do we still need this if we're expanding all stamped headers?

File: celery/canvas.py [hunk 3]
```diff
@@ -262,7 +262,7 @@ class Signature(dict):
     _app = _type = None
     # The following fields must not be changed during freezing/merging because
     # to do so would disrupt completion of parent tasks
-    _IMMUTABLE_OPTIONS = {"group_id", "groups"}
+    _IMMUTABLE_OPTIONS = {"group_id", "groups", "stamps"}
```
Comment (thedrow) [inline celery/canvas.py:265]:
```suggestion
    _IMMUTABLE_OPTIONS = {"group_id", "stamps"}
```

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/canvas.py (hunk 3 above)
Comment (odobosevych) [reply celery/canvas.py]: removed

File: celery/app/task.py (hunk 2 above)
Comment (dobosevych) [reply celery/app/task.py]: removed, everything is expanding now

**Comment by dobosevych** [comment]
Changed this PR to master branch and will keep that, as a main branch of development

**Comment by auvipy** [comment]
OK cool

**Comment by thedrow** [comment]
@dobosevych Please edit the PR's title and add a description.

File: celery/backends/base.py [hunk 4]
```diff
@@ -488,8 +488,11 @@ def _get_result_meta(self, result,
                     'retries': getattr(request, 'retries', None),
                     'queue': request.delivery_info.get('routing_key')
                     if hasattr(request, 'delivery_info') and
-                    request.delivery_info else None
+                    request.delivery_info else None,
                 }
+                if hasattr(request, 'stamps') and request.stamps:
```
Comment (thedrow) [inline celery/backends/base.py:493]:
```suggestion
                if getattr(request, 'stamps'):
```

**Comment by dobosevych** [comment]
> @dobosevych Please edit the PR's title and add a description.

Ok. I will improve that in the evening

File: celery/canvas.py [hunk 5]
```diff
@@ -725,7 +910,7 @@ def prepare_steps(self, args, kwargs, tasks,
                 task = from_dict(task, app=app)
             if isinstance(task, group):
                 # when groups are nested, they are unrolled - all tasks within
-                # groups within groups should be called in parallel
+                # groups should be called in parallel
```
Comment (thedrow) [inline celery/canvas.py:913]: Please revert this or describe why you think this comment is inaccurate.

File: celery/canvas.py [hunk 6]
```diff
@@ -1219,7 +1424,10 @@ def _apply_tasks(self, tasks, producer=None, app=None, p=None,
     def _freeze_gid(self, options):
         # remove task_id and use that as the group_id,
         # if we don't remove it then every task will have the same id...
-        options = dict(self.options, **options)
+        options = {**self.options, **{
+                k: v for k, v in options.items()
+                if k not in self._IMMUTABLE_OPTIONS or k not in self.options
+            }}
```
Comment (thedrow) [inline celery/canvas.py:1430]: This is a duplicate of https://github.com/celery/celery/pull/7384/files#diff-3a80ff45da16a11b96e26a63973d7d490187a68ddc1949e2dfd7fd090b208841R381-R387 

Please extract to a helper function

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/canvas.py (hunk 5 above)
Comment (odobosevych) [reply celery/canvas.py:913]: reverted

File: docs/userguide/canvas.rst [hunk 7]
```diff
@@ -1122,3 +1122,53 @@ of one:
 
 This means that the first task will have a countdown of one second, the second
 task a countdown of two seconds, and so on.
+
+Stamping
+==============
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1127]:
```suggestion
Stamping
=======
```

File: docs/userguide/canvas.rst [hunk 8]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+    >>> sig1_res._get_task_meta()['stamp']
+    ['your_custom_stamp']
+
+will initialize a group ``g`` and mark its components with stamp ``your_custom_stamp``.
+
+You need to set ``result_extended = True`` for the usage of stamping API.
+
+
+Group stamping
+------
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1148]:
```suggestion
Group Stamping
----------------
```

File: docs/userguide/canvas.rst [hunk 9]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+In addition to the default group stamping, we can also stamp
+canvas with custom marks, as shown in the example.
+
+Custom stamping
+------
+If more complex stamping logic is required, it is possible
+to implement custom stamping behavior based on the Visitor
+pattern. The class that implements this custom logic must
+inherit ``VisitorStamping`` and implement appropriate methods.
+
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1174]: We need to elaborate and have examples here.

File: docs/userguide/canvas.rst [hunk 10]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+    >>> g.apply_async()
+
+the header of task sig1 will store the stamp groups with g.id.
+In the case of nested groups, the order of the stamps corresponds
+to the nesting level. The group stamping is idempotent;
+the task cannot be stamped twice with the same group id.
+
+Canvas stamping
+------
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1164]:
```suggestion
Canvas Stamping
-----------------
```

File: docs/userguide/canvas.rst [hunk 11]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+In addition to the default group stamping, we can also stamp
+canvas with custom marks, as shown in the example.
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1166]:
```suggestion
canvas with custom stamps, as shown in the example.
```

**Review by thedrow — CHANGES_REQUESTED** [review]
I'm not satisfied with the documentation yet. There are a few RST syntax errors as well.

File: docs/userguide/canvas.rst [hunk 12]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+    >>> g.stamp(stamp='your_custom_stamp')
+    >>> res = g1.apply_async()
+    >>> res.get(timeout=TIMEOUT)
+    [4, 6]
+    >>> sig1_res._get_task_meta()['stamp']
+    ['your_custom_stamp']
+
+will initialize a group ``g`` and mark its components with stamp ``your_custom_stamp``.
+
+You need to set ``result_extended = True`` for the usage of stamping API.
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1144]: This is a setting and we can refer to it using a directive.

File: docs/userguide/canvas.rst [hunk 13]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+    >>> g.stamp(stamp='your_custom_stamp')
+    >>> res = g1.apply_async()
+    >>> res.get(timeout=TIMEOUT)
+    [4, 6]
+    >>> sig1_res._get_task_meta()['stamp']
+    ['your_custom_stamp']
+
+will initialize a group ``g`` and mark its components with stamp ``your_custom_stamp``.
+
+For the usage of stamping API to use this feature, you need to set up the :ref:`configuration <configuration:result_extended>` ``CELERY_RESULT_EXTENDED = True``
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1144]: There's a custom directive for referencing settings. See https://github.com/celery/celery/blob/c506f45926b40c041bfbe5147e3a3e59a9751435/docs/userguide/optimizing.rst#L58 for example.

File: docs/userguide/canvas.rst [hunk 14]
```diff
@@ -1122,3 +1122,53 @@ of one:
...
+
+    >>> sig1 = add.si(2, 2)
+    >>> sig1_res = sig1.freeze()
+    >>> g = group(sig1, add.si(3, 3))
+    >>> g.stamp(stamp='your_custom_stamp')
+    >>> res = g1.apply_async()
+    >>> res.get(timeout=TIMEOUT)
+    [4, 6]
+    >>> sig1_res._get_task_meta()['stamp']
+    ['your_custom_stamp']
```
Comment (thedrow) [inline docs/userguide/canvas.rst:1140]: We should demonstrate the default behavior of stamping two values on the same key.

**Review by thedrow — APPROVED** [review]

**Review by auvipy — DISMISSED** [review]
need to investigate the following failure

  -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
  =========================== short test summary info ============================
  SUBFAIL t/integration/test_canvas.py::test_chord::test_mutable_errback_called_by_chord_from_group_fail_multiple[errback_old_style]

**Comment by auvipy** [comment]
Ok

**Comment by odobosevych** [comment]
Ok. I will will do one more iteration

Best regards, Oles Dobosevych

**Pull request merged**